expression/builtin_miscellaneous_vec.go

```diff
@@ -281,11 +284,80 @@ func (b *builtinNameConstTimeSig) vecEvalTime(input *chunk.Chunk, result *chunk.
 }
 
 func (b *builtinSleepSig) vectorized() bool {
-	return false
+	return true
 }
 
+// vecEvalInt evals a builtinSleepSig in a vectorized manner.
+// See https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_sleep
 func (b *builtinSleepSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETReal, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+
+	err = b.args[0].VecEvalReal(b.ctx, input, buf)
+	if err != nil {
+		return err
+	}
+
+	result.ResizeInt64(n, false)
+	i64s := result.Int64s()
+
+	for i := 0; i < n; i++ {
+		isNull := buf.IsNull(i)
+		val := buf.GetFloat64(i)
+		sessVars := b.ctx.GetSessionVars()
+		if isNull {
+			if sessVars.StrictSQLMode {
+				return errIncorrectArgs.GenWithStackByArgs("sleep")
+			}
+			return nil
+		}
+		// processing argument is negative
+		if val < 0 {
+			if sessVars.StrictSQLMode {
+				return errIncorrectArgs.GenWithStackByArgs("sleep")
+			}
+			return nil
+		}
+
+		if val > math.MaxFloat64/float64(time.Second.Nanoseconds()) {
+			return errIncorrectArgs.GenWithStackByArgs("sleep")
+		}
+
+		err := doSleep(val, sessVars)
+		if err != nil {
+			for j := i; j < n; j++ {
+				i64s[j] = 1
+			}
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func doSleep(secs float64, sessVars *variable.SessionVars) error {
+	dur := time.Duration(secs * float64(time.Second.Nanoseconds()))
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
+	start := time.Now()
+	finish := false
+	for !finish {
+		select {
+		case now := <-ticker.C:
+			if atomic.CompareAndSwapUint32(&sessVars.Killed, 1, 0) {
+				return errors.New("session is killed")
+			}
+
+			if now.Sub(start) > dur {
+				finish = true
```

Comment:
`finish` can be eliminated here.
```suggestion
				return false
```